### PR TITLE
More Cross Compilation Work

### PR DESCRIPTION
**Pull request by eeeebbbbrrrr**
I'll come back to this tomorrow...

### CODE REVIEW

**Comment by eeeebbbbrrrr** [comment]
I'm going to work on the PR description above and then also work on the changes to README.md to talk about cross-compiling, "trusted", and the `trusted-pgx` crate.  Neither of those should hold up a review and a merge.

I think this is in good (enough) shape and passes my smoke testing locally.

File: README.md [hunk 1]
```diff
@@ -1,73 +1,93 @@
...
 
-Further usage requires setting the `PLRUST_TARGET` to the appropriate tuple, as this is still an experimental feature,
-and so is not currently set by default. Passing `--features target_postgrestd` to `cargo` also works. These are subtly different:
+`postgrestd` and the `trusted` feature flag are **not** supported on other platforms.  As such, PL/Rust cannot be 
+considered fully trusted on those platforms.
 
-- `PLRUST_TARGET` is evaluated when the PL/Rust handler builds a function
-- `--features target_postgrestd` is evaluated when the PL/Rust handler is built
-- `PLRUST_TARGET` prevails over `target_postgrestd` if both are set
+## TODO:  PL/Rust is also Cross Compiler
```
Comment (thomcc) [inline README.md:53]:
```suggestion
## TODO:  PL/Rust is also a Cross Compiler
```

File: plrust/src/error.rs [hunk 2]
```diff
@@ -32,6 +30,6 @@ pub(crate) enum PlRustError {
     ParsingCodeBlock(syn::Error),
     #[error("Parsing error at span `{:?}`", .0.span())]
     Parse(#[from] syn::Error),
-    #[error("No plrust.plrust_proc entry for ({0}, {1})")]
-    NoProcEntry(pgx::pg_sys::Oid, CompilationTarget),
+    #[error("plrust.plrust_proc.so bytes are NULL")]
+    NullPlRustProcSharedLibraryBytes,
```
Comment (workingjubilee) [inline plrust/src/error.rs:34]: Does this mean they are all `\0`s or that it is a pointer to address 0x0?

**Review by thomcc — APPROVED** [review]
Looks great! I don't have a ton of context on the plrust crate's internals so it's mostly just signing off that the parts I do understand look right, and they do!

File: plrust/src/lib.rs [hunk 3]
```diff
@@ -10,6 +10,27 @@ Use of this source code is governed by the PostgreSQL license that can be found
...
+        any(target_arch = "x86_64", target_arch = "aarch64")
+    ))
+))]
+compile_error!("This platform does not support the 'trusted' version of plrust");
+
+cfg_if::cfg_if! {
+    if #[cfg(feature = "trusted")] {
+        /// We are a trusted language handler.  This will cause plrust user functions to be compiled
+        /// with `postgrestd`
+        pub(crate) const TRUSTED:bool = true;
```
Comment (workingjubilee) [inline plrust/src/lib.rs:26]: extremely minor and tedious nit, nonetheless,
```suggestion
        pub(crate) const TRUSTED: bool = true;
```

File: plrust/src/lib.rs [hunk 4]
```diff
@@ -10,6 +10,27 @@ Use of this source code is governed by the PostgreSQL license that can be found
...
+    } else {
+        /// We are NOT a trusted language handler.  This will cause plrust user functions to be compiled
+        /// with the standard Rust `std`
+        pub(crate) const TRUSTED:bool = false;
```
Comment (workingjubilee) [inline plrust/src/lib.rs:30]:
```suggestion
        pub(crate) const TRUSTED: bool = false;
```

File: plrust/src/target.rs [hunk 5]
```diff
@@ -104,27 +113,83 @@ impl CompilationTarget {
...
-                        Ok(host::target_tuple().into())
-                    }
-                }
-            }
-            Err(env::VarError::NotUnicode(s)) => Err(TargetErr::InvalidSpec(s)),
+#[derive(Clone, Copy, PartialEq, Hash, Debug)]
+pub(crate) enum CrossCompilationTarget {
+    X86_64 = 1,
+    Aarch64 = 2,
+}
```
Comment (workingjubilee) [inline plrust/src/target.rs:120]: Why are we assigning values to these variants? We usually do that in enums we want to be FFI compatible, but there is no repr attribute here.

File: plrust/src/error.rs (hunk 2 above)
Comment (eeeebbbbrrrr) [reply plrust/src/error.rs:34]: It means the datum is an SQL NULL.

File: plrust/src/target.rs (hunk 5 above)
Comment (eeeebbbbrrrr) [reply plrust/src/target.rs:120]: Oversight from an earlier incarnation. Will fix. Good catch.

File: plrust/src/error.rs (hunk 2 above)
Comment (eeeebbbbrrrr) [reply plrust/src/error.rs:34]: The column itself is declared NOT NULL so this would only happen if Postgres broke

Comment (workingjubilee) [reply plrust/src/error.rs:34]: Ah okay.
In such a case we should probably simply panic, but
```suggestion
    #[error("plrust.plrust_proc.so is an SQL NULL despite being NOT NULL")]
    NullPlRustProcSharedLibraryBytes,
```

File: plrust/src/target.rs (hunk 5 above)
Comment (workingjubilee) [reply plrust/src/target.rs]: Yeah, I'm not so concerned about doing so as much as I was just checking that it wasn't being used in an odd way elsewhere that escaped my notice.

**Review by workingjubilee — APPROVED** [review]
This looks correct. Thank you.

**Comment by eeeebbbbrrrr** [comment]
> > test tests::tests::pg_plrust_one_sleepy_boi ... FAILED
> 
> Between all the PRs being merged, the assert somehow made its return...?

It seems to be intermittent. Which is another level of awesome.

**Comment by workingjubilee** [comment]
I'm going to merge this over the error and address the issue in followup if needs be.

**Pull request merged**